### PR TITLE
Tweak type alias logics

### PR DESCRIPTION
**Pull request by kc0506**
## Change Summary

Put type alias logic after `from_property`.

## Checklist

* [ ] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

**Comment by sydney-runkle** [comment]
@Viicos, could you take a look at this one? You've had your head in this space recently...

File: pydantic/_internal/_generate_schema.py [hunk 1]
```diff
@@ -1375,13 +1376,13 @@ def _union_schema(self, union_type: Any) -> core_schema.CoreSchema:
 
     def _type_alias_type_schema(
         self,
-        obj: TypeAliasType,
+        obj: TypeAliasType | GenericAlias,
```
Comment (Viicos) [inline pydantic/_internal/_generate_schema.py:1379]: `GenericAlias` and `typing._GenericAlias` act as proxy objects in some way, I don't think we want to annotate `obj` like so here.

File: pydantic/_internal/_generate_schema.py [hunk 2]
```diff
@@ -1375,13 +1376,13 @@ def _union_schema(self, union_type: Any) -> core_schema.CoreSchema:
...
     ) -> CoreSchema:
         with self.defs.get_schema_or_ref(obj) as (ref, maybe_schema):
             if maybe_schema is not None:
                 return maybe_schema
 
-            origin = get_origin(obj) or obj
+            origin: TypeAliasType = get_origin(obj) or cast(TypeAliasType, obj)
```
Comment (Viicos) [inline pydantic/_internal/_generate_schema.py:1385]: Probably best to just do:

```suggestion
            origin: TypeAliasType = get_origin(obj) or obj
```

File: pydantic/_internal/_generate_schema.py [hunk 3]
```diff
@@ -1039,7 +1038,9 @@ def _match_generic_type(self, obj: Any, origin: Any) -> CoreSchema:  # noqa: C90
         if from_property is not None:
             return from_property
 
-        if _typing_extra.origin_is_union(origin):
+        if isinstance(origin, TypeAliasType):
+            return self._type_alias_type_schema(obj)
```
Comment (Viicos) [inline pydantic/_internal/_generate_schema.py:1042]: Why was it moved?

File: pydantic/_internal/_generate_schema.py (hunk 1 above)
Comment (kc0506) [reply pydantic/_internal/_generate_schema.py:1379]: Right, I forgot about that.

File: pydantic/_internal/_generate_schema.py (hunk 3 above)
Comment (kc0506) [reply pydantic/_internal/_generate_schema.py:1042]: I think without a special reason (e.g. `dataclass` below), these matching clauses should better be put together. Looks like when `origin` is of type `TypeAliasType`, `from_property` is always `None` (at least in the current tests), so moving it does not have any actual effect.

**Review by Viicos — APPROVED** [review]

**Pull request merged**